From afa8ff4c5135b955d195eeb17ca7d8d612d6724e Mon Sep 17 00:00:00 2001
From: Dave Taht <dave.taht@bufferbloat.net>
Date: Fri, 10 Feb 2012 15:29:12 +0000
Subject: [PATCH 2/2] Hardlinks can fail across union mounts, use rename on
 conf file

Hardlinks fails on UNION MOUNTs and similar filesystems if the
config file is still on the RO layer. Hardlinks across layers
will not work and cause quagga to fail saving the configuration.

Rename is atomic and 'does the right thing'
---
 lib/command.c |   12 +++---------
 1 files changed, 3 insertions(+), 9 deletions(-)

diff --git a/lib/command.c b/lib/command.c
index e62a7a7..fc76f35 100644
--- a/lib/command.c
+++ b/lib/command.c
@@ -2601,20 +2601,14 @@ DEFUN (config_write_file,
 		 VTY_NEWLINE);
         goto finished;
       }
-  if (link (config_file, config_file_sav) != 0)
+  if (rename (config_file, config_file_sav) != 0)
     {
-      vty_out (vty, "Can't backup old configuration file %s.%s", config_file_sav,
+      vty_out (vty, "Can't rename old configuration file %s.%s", config_file_sav,
 	        VTY_NEWLINE);
       goto finished;
     }
   sync ();
-  if (unlink (config_file) != 0)
-    {
-      vty_out (vty, "Can't unlink configuration file %s.%s", config_file,
-	        VTY_NEWLINE);
-      goto finished;
-    }
-  if (link (config_file_tmp, config_file) != 0)
+  if (rename (config_file_tmp, config_file) != 0)
     {
       vty_out (vty, "Can't save configuration file %s.%s", config_file,
 	       VTY_NEWLINE);
-- 
1.7.5.4

